core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -732,18 +742,20 @@ class GroupMetadataManager(brokerId: Int,
     * @param groups Groups whose metadata are to be cleaned up
     * @param selector A function that implements deletion of (all or part of) group offsets. This function is called while
     *                 a group lock is held, therefore there is no need for the caller to also obtain a group lock.
+    * @param onGroupUnloaded A function that handles group unload.
     * @return The cumulative number of offsets removed
     */
-  def cleanupGroupMetadata(groups: Iterable[GroupMetadata], selector: GroupMetadata => Map[TopicPartition, OffsetAndMetadata]): Int = {
+  def cleanupGroupMetadata(groups: Iterable[GroupMetadata],
+                           selector: GroupMetadata => Map[TopicPartition, OffsetAndMetadata],
+                           onGroupUnloaded: GroupMetadata => Unit): Int = {
     var offsetsRemoved = 0
 
     groups.foreach { group =>
       val groupId = group.groupId
       val (removedOffsets, groupIsDead, generation) = group.inLock {
         val removedOffsets = selector(group)
         if (group.is(Empty) && !group.hasOffsets) {
-          info(s"Group $groupId transitioned to Dead in generation ${group.generationId}")
-          group.transitionTo(Dead)
+          removeGroup(group, onGroupUnloaded)
```

Comment:
I think we cannot do this directly here, since we rely on line 770 below to write tombstone records. Copying my thoughts from the other PR here: 

```
A group can only be removed from the metadata cache in three ways:

a) a group must first become empty (but it would not be Dead yet). And then after the offset expiry time the cleanup thread may finally be able to remove its offsets, and only when group.is(Empty) && !group.hasOffsets it will then transit to Dead. And then this group will be removed.

b) Due to offset-topic partition leader migration, the group would be "unloaded". In this case it will transit to Dead immediately and then be removed.

c) A delete-group API request is received and the group is indeed empty already, which it will to transit to Dead immediately. And later on the same patten of cleanup thread in a) will eventually remove it from cache.
```

In these two situations, a) and c) are related to "deleting a group", i.e. the broker still owns the group's id range since it still holds the leader partition and now needs to remove it since the group no longer exists, in this case it is responsible for writing the tombstone record so that other follower replicas would be notified. b) are related to "unloading a group", i.e. the broker no longer owns the group since the leader partition has migrated out, in this case it does not need to write the tombstone since it is other broker's responsibility now.

Reply:
I think the only case we should consider improving is case c) above, and this PR seems too much over-complicated. And now that I think about this a bit longer, I think it has a rationale to not do the physical tombstone since it would tie up one processor thread, whereas it is more efficient to let the scheduler thread to do the actual removal which incurs writing the tombstone physically on logs as well.

What we only need to improve then, is that in handleDeleteGroups (i.e. case c). Currently the semantics is that "even after the delete-group is received, we still let its committed offsets to stay until expired, and then removed the group and offsets" --> note that we need to write tombstones for both deleted offsets as well as deleted groups. However when handling offsetCommit request if the group is already in Dead state we return INVALID_OFFSET anyways, so there's no point maintaining the offsets anymore after handleDeleteGroups. However we cannot directly remove the offsets from cache since we still need to write the tombstones. 

So what we can do is to just set `expireTimestamp` directly to now so that it can soon be reaped by the scheduler thread. And I feel that's the only optimization we'd need for now.